### DISCUSSION TITLE
Environment variable setup issue for self-hosted runner

### DISCUSSION THREAD

**Discussion by chenrui333** [community/community#26521] [Actions]
<p>I have deployed self-hosted runners (confguration in <a href="https://github.com/chenrui333/actions-runner-controller/blob/master/runner/Dockerfile" rel="nofollow noopener">here</a>).</p>
<p>But when I tried to run it for our internal android project, it complains about missing the ANDROID_HOME environemnt variable. But it does exist in the /etc/environemnt though:</p>
<pre><code>runner@meetup-android-runner-nb7zv-7wz42:/$ cat /etc/environment
PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games"
JAVA_HOME_8_X64=/usr/lib/jvm/zulu-8-azure-amd64
JAVA_HOME=/usr/lib/jvm/zulu-8-azure-amd64
JAVA_TOOL_OPTIONS=-Dfile.encoding=UTF8
ANT_HOME=/usr/share/ant
ANDROID_SDK_ROOT=/usr/local/lib/android/sdk
ANDROID_HOME=/usr/local/lib/android/sdk
</code></pre>
<p>So the temporary solution for us is to manually setup environemtn variable in the Github Action workflows like this:</p>
<pre><code>steps:
      - name: Set runner-specific environment variables
        shell: bash
        run: |
          echo "::set-env name=ANDROID_HOME::/usr/local/lib/android/sdk"
          echo "::set-env name=ANDROID_SDK_ROOT=/usr/local/lib/android/sdk"
</code></pre>
<p>I wonder why the /etc/environment did not get consumed by the container and what would be the better appraoch to solve this problem. </p>
<p>Side note, the runner is built on top of actions/runner v2.168.0</p>

**Comment by rbellamy** [comment, marked as answer]
<p>It’s because you launch the runner from your entrypoint, which isn’t a “login” shell. In order to leverage /etc/environment, you MUST have a login shell because that file is loaded by the “pam_env” module.</p>
<p>To fix your Dockerfile, you’ll need to execute the “run.sh” or “runsvc.sh” scripts from CMD, or set up your environment in a different manner.</p>
<p>Example:</p>
<p><a href="https://github.com/terradatum/github-runner" rel="nofollow noopener">https://github.com/terradatum/github-runner</a></p>

**Comment by chenrui333** [comment]
<p>Thanks for the response, I did take a look at your Dockerfile, do you mean I should update the mine to be something like this?</p>
<pre><code>COPY entrypoint.sh /sbin/entrypoint.sh
COPY supervisord.conf /etc/supervisor/conf.d/supervisord.conf
RUN chmod +x /sbin/entrypoint.sh \
  &amp;&amp; chmod 644 /etc/supervisor/conf.d/supervisord.conf
ENTRYPOINT ["/sbin/entrypoint.sh"]
CMD ["/usr/bin/supervisord", "-c", "/etc/supervisor/conf.d/supervisord.conf"]

...

USER runner:runner
</code></pre>
<p>I am less familiar with /etc/environment flow, can you be a little bit specific. Thanks for the help!</p>

**Comment by rbellamy** [comment]
<p>Bottom line: if you want to use /etc/environment you must understand how the entrypoint will or will not load a shell: <a href="https://docs.docker.com/engine/reference/builder/#shell-form-entrypoint-example" rel="nofollow noopener">https://docs.docker.com/engine/reference/builder/#shell-form-entrypoint-example</a>.</p>
<p>The advantage of using something like supervisord or tini is that these tools will deal with zombie reaping as PID 1: <a href="https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/" rel="nofollow noopener">https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/</a>.</p>
<p>Note that Docker uses tini when running a container with the “–init” flag: <a href="https://docs.docker.com/engine/reference/run/#specify-an-init-process" rel="nofollow noopener">https://docs.docker.com/engine/reference/run/#specify-an-init-process</a>.</p>

**Comment by chenrui333** [comment]
<p>I will read about this, thanks for all these documentation links.</p>
